File: python/src/baselode/extent.py
# SPDX-License-Identifier: GPL-3.0-or-later

# Copyright (C) 2026 Darkmine Pty Ltd

# This file is part of baselode.

# baselode is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# baselode is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with baselode.  If not, see <https://www.gnu.org/licenses/>.

import shapely.geometry

# Canonical default — GeoJSON / OGC convention.
DEFAULT_CRS = "EPSG:4326"


def _normalise_crs(crs):
    """Validate ``crs`` and return it in canonical form.

    Mirrors how geopandas accepts CRS values: anything
    ``pyproj.CRS.from_user_input`` understands (EPSG strings like
    ``"EPSG:4326"``, bare ints like ``4326``, WKT/proj strings, or a
    ``pyproj.CRS`` instance) is allowed. Anything else raises
    ``ValueError`` with a helpful message.

    The returned value is the canonical form:

    - ``"EPSG:N"`` (string) when the CRS has an EPSG code (the common case).
    - the parsed ``pyproj.CRS`` instance when it doesn't (e.g. custom WKT
      with no registered EPSG code) — preserves full fidelity.
    """
    if crs is None:
        raise ValueError(
            "Extent requires a CRS. Pass crs='EPSG:4326' (default) or any "
            "value pyproj.CRS.from_user_input accepts."
        )
    try:
        import pyproj
    except ImportError as exc:
        raise ValueError(
            f"Cannot validate CRS {crs!r} because pyproj is not installed. "
            "Install pyproj (it's a baselode hard dependency)."
        ) from exc
    try:
        parsed = pyproj.CRS.from_user_input(crs)
    except Exception as exc:
        raise ValueError(
            f"Invalid CRS {crs!r}: {exc}. Pass anything "
            "pyproj.CRS.from_user_input accepts (e.g. 'EPSG:4326', 28351, "
            "a WKT string, or a pyproj.CRS instance)."
        ) from exc
    epsg = parsed.to_epsg()
    if epsg is not None:
        return f"EPSG:{epsg}"
    return parsed


class Extent():

    def __init__(self, xmin=None, xmax=None, ymin=None, ymax=None, bbox=None, name=None, crs=DEFAULT_CRS):
        """
        Create an extent object, which is an axis-aligned bounding box with name and
        coordinate reference system (CRS).

        Pass either:
        @param bbox - the axis aligned bounding box as a shapely.geometry.box object
        OR
        @param xmin, xmax, ymin, ymax - the coordinates of the bounding box edges

        @param name - optional name for the extent
        @param crs - coordinate reference system. Anything ``pyproj.CRS.from_user_input``
            accepts works (e.g. ``"EPSG:4326"``, the bare int ``4326``, a WKT string,
            or a ``pyproj.CRS`` instance). Defaults to ``"EPSG:4326"``. Validated
            and normalised at construction; an invalid value raises ``ValueError``.
            See :meth:`set_crs` for the normalisation rules.
        """
        if bbox is None:
            self.bbox = shapely.geometry.box(xmin, ymin, xmax, ymax)
            self.set_minmax()
        else:
            self.bbox = bbox
            self.set_minmax()
        self.name = name
        self.set_crs(crs)

    def set_minmax(self):
        self.xmin, self.ymin, self.xmax, self.ymax = self.bbox.bounds

    def set_crs(self, crs):
        """Validate, normalise and store ``crs`` on this Extent.

        Accepts anything ``pyproj.CRS.from_user_input`` accepts. The stored
        value is the canonical form:

        - ``"EPSG:N"`` (string) when the CRS has an EPSG code.
        - the parsed ``pyproj.CRS`` instance when it doesn't.

        After ``set_crs``, ``self.crs`` always reads back in normalised form
        regardless of what was passed in (e.g. ``e.set_crs(4326)`` →
        ``e.crs == "EPSG:4326"``; ``e.set_crs("epsg:28351")`` →
        ``e.crs == "EPSG:28351"``).

        Returns ``self`` so calls are chainable.
        """
        self.crs = _normalise_crs(crs)
        return self

    def to_crs(self, target_crs):
        """Return a new Extent with corners reprojected to ``target_crs``.

        Modelled on ``geopandas.GeoDataFrame.to_crs``. ``target_crs``
        accepts anything ``pyproj.CRS.from_user_input`` accepts and is
        validated / normalised the same way :meth:`set_crs` is.

        For lat/lon use cases::

            wgs84 = extent.to_crs("EPSG:4326")
            min_lon, min_lat, max_lon, max_lat = (
                wgs84.xmin, wgs84.ymin, wgs84.xmax, wgs84.ymax,
            )

        If ``target_crs`` already matches this Extent's CRS, returns a
        fresh copy with the same bounds (no pyproj transform). Otherwise
        reprojects via ``pyproj.Transformer``; reprojection failures
        raise ``ValueError`` rather than silently passing through.
        """
        target = _normalise_crs(target_crs)
        if target == self.crs:
            return Extent(
                xmin=self.xmin, xmax=self.xmax,
                ymin=self.ymin, ymax=self.ymax,
                name=self.name, crs=target,
            )
        try:
            import pyproj
            transformer = pyproj.Transformer.from_crs(
                pyproj.CRS.from_user_input(self.crs),
                pyproj.CRS.from_user_input(target),
                always_xy=True,
            )
            xs, ys = transformer.transform(
                [self.xmin, self.xmax], [self.ymin, self.ymax],
            )
        except Exception as exc:
            raise ValueError(
                f"Failed to reproject extent from CRS {self.crs!r} to "
                f"{target!r}: {exc}"
            ) from exc
        return Extent(
            xmin=min(xs), xmax=max(xs),
            ymin=min(ys), ymax=max(ys),
            name=self.name, crs=target,
        )

    def get_folium_rectangle(self):
        """Return bounds formatted for folium.Rectangle: [[south, west], [north, east]]."""
        xmin, ymin, xmax, ymax = self.bbox.bounds
        return [[ymin, xmin], [ymax, xmax]]

    def center(self, lonlat=False):
        """Return the bbox center as (x, y) in the extent's CRS, or (lon, lat) if requested.

        When lonlat=True and the extent has a CRS attribute, the centroid is reprojected to EPSG:4326.
        """
        xmin, ymin, xmax, ymax = self.bbox.bounds
        cx, cy = (xmin + xmax) / 2.0, (ymin + ymax) / 2.0
        if not lonlat:
            return cx, cy
        try:
            import pyproj

            source_crs = pyproj.CRS.from_user_input(self.crs) if hasattr(self, "crs") else pyproj.CRS.from_epsg(4326)
            target_crs = pyproj.CRS.from_epsg(4326)
            transformer = pyproj.Transformer.from_crs(source_crs, target_crs, always_xy=True)
            lon, lat = transformer.transform(cx, cy)
            return lon, lat
        except Exception:
            # Fallback to raw center if reprojection fails
            return cx, cy

File: javascript/packages/baselode/src/extent/Extent.js
/*
 * Copyright (C) 2026 Darkmine Pty Ltd
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

import proj4 from 'proj4';

export const DEFAULT_EXTENT_CRS = 'EPSG:4326';

const EPSG_PATTERN = /^epsg\s*:\s*(\d+)$/i;
const BARE_EPSG_PATTERN = /^\d+$/;

function registerAustralianCrsDefinitions() {
  if (!proj4.defs('EPSG:4283')) {
    proj4.defs(
      'EPSG:4283',
      '+proj=longlat +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +no_defs +type=crs',
    );
  }
  for (let zone = 49; zone <= 58; zone += 1) {
    const code = `EPSG:${28300 + zone}`;
    if (!proj4.defs(code)) {
      proj4.defs(
        code,
        `+proj=utm +zone=${zone} +south +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs +type=crs`,
      );
    }
  }
}

registerAustralianCrsDefinitions();

function describe(value) {
  if (typeof value === 'string') return JSON.stringify(value);
  return String(value);
}

function normaliseCrs(crs) {
  if (crs === null || crs === undefined) {
    throw new TypeError(
      "Extent requires a CRS. Pass crs: 'EPSG:4326' (default), an EPSG number, or a proj/WKT definition.",
    );
  }

  let normalized;
  if (typeof crs === 'number') {
    if (!Number.isSafeInteger(crs) || crs <= 0) {
      throw new TypeError(`Invalid CRS ${describe(crs)}: EPSG numbers must be positive integers.`);
    }
    normalized = `EPSG:${crs}`;
  } else if (typeof crs === 'string') {
    const trimmed = crs.trim();
    if (!trimmed) {
      throw new TypeError('Invalid CRS "": CRS definitions must not be empty.');
    }
    const epsgMatch = trimmed.match(EPSG_PATTERN);
    if (epsgMatch) {
      normalized = `EPSG:${Number(epsgMatch[1])}`;
    } else if (BARE_EPSG_PATTERN.test(trimmed)) {
      normalized = `EPSG:${Number(trimmed)}`;
    } else {
      normalized = trimmed;
    }
  } else {
    throw new TypeError(
      `Invalid CRS ${describe(crs)}: pass an EPSG string/number or a proj/WKT definition.`,
    );
  }

  try {
    proj4(normalized);
  } catch (error) {
    const detail = error?.message ? `: ${error.message}` : '';
    throw new TypeError(
      `Invalid CRS ${describe(crs)}${detail}. The definition is not registered or supported by proj4.`,
      { cause: error },
    );
  }
  return normalized;
}

function finiteNumber(value, name) {
  if (typeof value !== 'number' || !Number.isFinite(value)) {
    throw new TypeError(`Extent ${name} must be a finite number; received ${describe(value)}.`);
  }
  return value;
}

function boundsFromBbox(bbox) {
  if (Array.isArray(bbox)) {
    if (bbox.length !== 4) {
      throw new TypeError('Extent bbox arrays must contain [west, south, east, north].');
    }
    return {
      xmin: bbox[0],
      ymin: bbox[1],
      xmax: bbox[2],
      ymax: bbox[3],
    };
  }
  if (bbox && typeof bbox === 'object') {
    return {
      xmin: bbox.west,
      ymin: bbox.south,
      xmax: bbox.east,
      ymax: bbox.north,
    };
  }
  throw new TypeError(
    'Extent bbox must be [west, south, east, north] or { west, south, east, north }.',
  );
}

function validateBounds(bounds) {
  const normalized = {
    xmin: finiteNumber(bounds.xmin, 'xmin'),
    ymin: finiteNumber(bounds.ymin, 'ymin'),
    xmax: finiteNumber(bounds.xmax, 'xmax'),
    ymax: finiteNumber(bounds.ymax, 'ymax'),
  };
  if (normalized.xmin > normalized.xmax) {
    throw new RangeError('Extent xmin must be less than or equal to xmax.');
  }
  if (normalized.ymin > normalized.ymax) {
    throw new RangeError('Extent ymin must be less than or equal to ymax.');
  }
  return normalized;
}

function transformedBounds(sourceCrs, targetCrs, points) {
  try {
    const transformed = points.map((point) => proj4(sourceCrs, targetCrs, point));
    const xs = transformed.map(([x]) => finiteNumber(x, 'transformed x'));
    const ys = transformed.map(([, y]) => finiteNumber(y, 'transformed y'));
    return {
      xmin: Math.min(...xs),
      ymin: Math.min(...ys),
      xmax: Math.max(...xs),
      ymax: Math.max(...ys),
    };
  } catch (error) {
    throw new Error(
      `Failed to reproject extent from CRS ${describe(sourceCrs)} to ${describe(targetCrs)}: ${error?.message || error}`,
      { cause: error },
    );
  }
}

/**
 * Axis-aligned bounding box with an explicit coordinate reference system.
 */
export class Extent {
  constructor(options = {}) {
    if (!options || typeof options !== 'object' || Array.isArray(options)) {
      throw new TypeError('Extent constructor expects an options object.');
    }
    const hasBbox = options.bbox !== undefined;
    const hasBounds = ['xmin', 'ymin', 'xmax', 'ymax']
      .some((property) => options[property] !== undefined);
    if (hasBbox && hasBounds) {
      throw new TypeError('Extent accepts either bbox or xmin/ymin/xmax/ymax, not both.');
    }
    const bounds = validateBounds(hasBbox ? boundsFromBbox(options.bbox) : options);
    const name = options.name ?? null;
    if (name !== null && typeof name !== 'string') {
      throw new TypeError('Extent name must be a string or null.');
    }
    const crs = normaliseCrs(options.crs ?? DEFAULT_EXTENT_CRS);

    Object.defineProperties(this, {
      xmin: { value: bounds.xmin, enumerable: true },
      ymin: { value: bounds.ymin, enumerable: true },
      xmax: { value: bounds.xmax, enumerable: true },
      ymax: { value: bounds.ymax, enumerable: true },
      name: { value: name, enumerable: true },
      crs: { value: crs, enumerable: true, writable: true },
    });
  }

  static fromBbox(bbox, crs = DEFAULT_EXTENT_CRS, name = null) {
    return new Extent({ bbox, crs, name });
  }

  static fromFeature(feature, crs = DEFAULT_EXTENT_CRS, name = null) {
    if (!feature || feature.type !== 'Feature' || feature.geometry?.type !== 'Polygon') {
      throw new TypeError('Extent.fromFeature expects a GeoJSON Polygon Feature.');
    }
    const positions = feature.geometry.coordinates.flat();
    if (!positions.length) {
      throw new TypeError('Extent.fromFeature cannot derive bounds from an empty Polygon.');
    }
    const xs = positions.map((position) => finiteNumber(position?.[0], 'polygon x'));
    const ys = positions.map((position) => finiteNumber(position?.[1], 'polygon y'));
    return new Extent({
      xmin: Math.min(...xs),
      ymin: Math.min(...ys),
      xmax: Math.max(...xs),
      ymax: Math.max(...ys),
      crs,
      name,
    });
  }

  setCrs(crs) {
    this.crs = normaliseCrs(crs);
    return this;
  }

  toCrs(targetCrs) {
    const target = normaliseCrs(targetCrs);
    if (target === this.crs) {
      return new Extent({
        xmin: this.xmin,
        ymin: this.ymin,
        xmax: this.xmax,
        ymax: this.ymax,
        name: this.name,
        crs: target,
      });
    }
    const bounds = transformedBounds(this.crs, target, [
      [this.xmin, this.ymin],
      [this.xmax, this.ymin],
      [this.xmax, this.ymax],
      [this.xmin, this.ymax],
    ]);
    return new Extent({ ...bounds, name: this.name, crs: target });
  }

  center({ lonlat = false } = {}) {
    const point = [
      (this.xmin + this.xmax) / 2,
      (this.ymin + this.ymax) / 2,
    ];
    if (!lonlat || this.crs === DEFAULT_EXTENT_CRS) return point;
    try {
      return proj4(this.crs, DEFAULT_EXTENT_CRS, point);
    } catch (error) {
      throw new Error(
        `Failed to reproject extent center from CRS ${describe(this.crs)} to ${describe(DEFAULT_EXTENT_CRS)}: ${error?.message || error}`,
        { cause: error },
      );
    }
  }

  getFoliumRectangle() {
    return [[this.ymin, this.xmin], [this.ymax, this.xmax]];
  }

  toPolygon() {
    return {
      type: 'Polygon',
      coordinates: [[
        [this.xmin, this.ymin],
        [this.xmax, this.ymin],
        [this.xmax, this.ymax],
        [this.xmin, this.ymax],
        [this.xmin, this.ymin],
      ]],
    };
  }

  toFeature(properties = {}) {
    if (properties !== null && (typeof properties !== 'object' || Array.isArray(properties))) {
      throw new TypeError('Extent feature properties must be an object or null.');
    }
    return {
      type: 'Feature',
      properties,
      geometry: this.toPolygon(),
    };
  }

  toFeatureCollection(properties = {}) {
    return {
      type: 'FeatureCollection',
      features: [this.toFeature(properties)],
    };
  }
}
